src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -200,6 +202,59 @@ internal static IEnumerable<ReplicationBatchItem> GetReplicationItems(DocumentDa
             }
         }
 
+        private class ReplicationBatchState : PulsedEnumerationState<ReplicationBatchItem>
+        {
+            private ReplicationDocumentSender _parent;
+            private ReplicationState _replicationState;
+            public bool WasInterrupted { get; private set; }
+            public long Next { get; private set; }
+
+            public ReplicationBatchState(DocumentsOperationContext context, Size pulseLimit, ReplicationDocumentSender parent, ReplicationState replicationState, long next) : base(context, pulseLimit)
+            {
+                _parent = parent;
+                _replicationState = replicationState;
+                WasInterrupted = false;
+                Next = next;
+            }
+
+            public override void OnMoveNext(ReplicationBatchItem current)
+            {
+                _parent._parent.ForTestingPurposes?.OnDocumentSenderFetchNewItem?.Invoke();
+                _parent._parent.CancellationToken.ThrowIfCancellationRequested();
+                _parent.AssertNoLegacyReplicationViolation(current);
+
+                if (_replicationState.LastTransactionMarker != current.TransactionMarker)
+                {
+                    _replicationState.Item = current;
+                    long next = Next;
+                    if (_parent.CanContinueBatch(_replicationState, ref next) == false)
+                    {
+                        WasInterrupted = true;
+                        return;
+                    }
+
+                    Next = next;
+                    _replicationState.LastTransactionMarker = current.TransactionMarker;
+                }
+            }
+
+            public override bool ShouldPulseTransaction()
+            {
+                if (_replicationState.NumberOfItemsSent == 0)
+                {
+                    var size = Context.Transaction.InnerTransaction.LowLevelTransaction.GetTotal32BitsMappedSize() +
+                               Context.Transaction.InnerTransaction.LowLevelTransaction.AdditionalMemoryUsageSize;
+                
+                    if (size >= PulseLimit)
+                    {
+                        return true;
+                    }
+                }
```

Comment:
we probably should call the `base` here

Reply:
I tried but there's still OOM if I do that..

Reply:
we need to use the `base` here, it is basically same code

Reply:
Done